Restore member-of-union expressions in trace output

tautschnig opened the pull request:
byte_extract operations precisely capture the semantics, but aren't
meaningful to users.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented:
It seems a bit odd to do the 'restoring' in the output of a trace; wouldn't it be better to do that when that is built?

tautschnig committed:
Restore member-of-union expressions in trace output

byte_extract operations precisely capture the semantics, but aren't
meaningful to users.

tautschnig commented:
> It seems a bit odd to do the 'restoring' in the output of a trace; wouldn't it be better to do that when that is built?

Indeed! I have reworked this, the patch is much smaller now.

kroening commented on src/goto-programs/goto_trace.h:
```diff
@@ -117,6 +117,7 @@ class goto_trace_stept
   bool cond_value;
   exprt cond_expr;
+  exprt original_condition;
 
```
The difference between `original_condition` and `cond_expr` is non-obvious, and does deserve a comment.
Do we really need both?

tautschnig replied:
I was tempted to just re-purpose `cond_expr`, but it turns out it is actually used by `goto-checker/bmc_util.cpp` (I believe that is the only use). `cond_expr` is the SSA-renamed expression, while we'd want to provide an expression that is meaningful to the user. So maybe adding a comment is the right approach?
